File: java/src/test/java/acceptance/steps/SubmarineSteps.java
package acceptance.steps;

import domain.Submarine;
import io.cucumber.java.en.And;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import static acceptance.Utils.FileUtils.getInputAsSeparatedLines;
import static java.util.Arrays.stream;
import static org.assertj.core.api.Assertions.assertThat;

public class SubmarineSteps {
    private final Submarine submarine;

    public SubmarineSteps() {
        submarine = new Submarine();
    }

    private static int calculateFinalValue(Submarine submarine) {
        return submarine.getPosition() * submarine.getDepth();
    }

    @When("^submarine receives command (.*)")
    public void receivesCommand(String command) {
        submarine.executeCommand(command);
    }

    @When("^submarine receives all commands from file (.*)$")
    public void receivesAllCommandsFromResourceFile(String fileName) {
        stream(getInputAsSeparatedLines(fileName))
                .forEach(submarine::executeCommand);
    }
    
    @Then("^submarine depth should be (\\d+)$")
    public void depthShouldBe(int expectedDepth) {
        assertThat(submarine.getDepth())
                .isEqualTo(expectedDepth);
    }

    @And("^submarine position should be (\\d+)$")
    public void positionShouldBe(int expectedPosition) {
        assertThat(submarine.getPosition())
                .isEqualTo(expectedPosition);
    }

    @And("^submarine aim should be (\\d+)$")
    public void aimShouldBe(int expectedAim) {
        assertThat(submarine.getAim())
                .isEqualTo(expectedAim);
    }

    @And("^submarine final value should be (\\d+)$")
    public void finalValueShouldBe(int expectedFinalValue) {
        assertThat(calculateFinalValue(submarine))
                .isEqualTo(expectedFinalValue);
    }
}

File: typescript/tests/specs/step-definitions/submarine-steps.ts
import {StepDefinitions} from "jest-cucumber";
import {Submarine} from "@submarine-domain/Submarine";


export const subamrineSteps: StepDefinitions = ({given, and, when, then}) => {
    let submarine = new Submarine()

    when(/submarine receives command (.*)$/, (command: string) => {
        submarine.executeCommand(command)
    })

    then(/^submarine depth should be (\d+)$/, (position: string) => {
        expect(submarine.depth()).toBe(parseInt(position))
    });

    then(/^submarine position should be (\d+)$/, (position: string) => {
        expect(submarine.position()).toBe(parseInt(position))
    });

    then(/^submarine aim should be (\d+)$/, (position: string) => {
        expect(submarine.aim()).toBe(parseInt(position))
    });

    then(/^submarine final value should be (\d+)$/, (position: string) => {
        expect(submarine.aim()).toBe(parseInt(position))
    });
}